### DISCUSSION TITLE
Q4-#1-You Only Cache Once: Decoder-Decoder Architectures for Language Models - Kyuhong Shim

### DISCUSSION THREAD

**Discussion by MarsJacobs** [efficient-ai-study/efficient-ai-study#21] [2024 Seminar]
:)

**Comment by lemonbuffer20** [comment]
발표 들어주신 모든 분들께 감사드립니다!

**Q1**. YOCO에서 전체 모델 단위에서 global KV cache를 공유하는 것이 핵심 아이디어라고 설명해주셨고, 추가로 소개해주신 다른 모델들로 미루어 보아 KV cache가 layer 단위로 공유되는 것이 가장 기본적인 구조인 것 같은데요. KV cache를 layer별로 다르게 두고 쓰는 것이 기본 구조로 자리 잡은 특별한 이유가 있었나요? (YOCO 이전에 global cache를 쓰는 시도가 왜 없었을지 궁금합니다)
**A1**. 우선, 대부분의 딥러닝 모델들은 layer 단위로 설계되었고, 동일 layer를 여러 층 쌓는 것으로 scale-up을 해 왔습니다. 이런 이유 때문에, 그리고 더 거슬러 올라가면 딥러닝의 기본 철학인 '서로 다른 층에서는 서로 다른 정보를 배운다'는 intuition 때문에, inter-layer 접근은 상대적으로 적었던 것 같습니다. 어찌 보면, 서로 다른 층인데 같은 KV를 쓰는게 딥러닝 모델 디자인 하는 입장에서는 더 파괴적인 혁신이라고 할 수도 있겠네요 :)

**Q2**. CLA는 연속하는 n개의 에텐션 헤드가 하나의 KV 캐시를 공유하는 방식인 것이라고 이해를 하였는데요, 혹시 기존의 CLA 기법 중에서는 n을 다양하게 사용한 경우가 있었을까요? 예를 들어 모델 하층부의 레이어는 2개의 헤드씩 묶어서 하나의 KV 캐시를 공유하게 하고, 중층부의 레이어는 3개, 하층부는 4개, ... 이런 식으로 헤드를 공유하는 레이어의 개수를 다양하게 활용한 접근은 없었던 것인지 궁금합니다.
**A2**. 묶는 층 수를 여러 가지로 조절해 본 연구들은 많지는 않습니다. 제 논문 (Understanding the Role of Self Attention for Efficient Speech Recognition)에도 비슷한 실험이 있기는 한데 (ex: 4+4+8 vs. 8+4+4) 음성 쪽이라 결과가 바로 여러 분야에 동일하게 적용되지는 않을 것 같습니다. MobileLLM 논문처럼, 보통 연속된 2개 층까지는 성능 하락이 크지 않다고 보고 많이 묶는 것 같습니다.

**Q3**. 해당 논문이 50/50으로 나눈 이유가 더 명확했으면 좋겠다라는 생각이 들었고, 각 파트가 하드웨어적인 오버헤드가 어떤식으로 다른지 궁금합니다. 
**A3-1**. 50/50은 최적 셋업을 찾았다기보다는 논문의 컨셉을 보여주기 위해 단순화한 거라고 생각하고 있습니다. 아마 하드웨어 따라, 데이터 따라, 최적 위치는 조금씩 달라질 것 같긴 합니다.
**A3-2**. 제가 직접 구현해 본 건 아니지만 제 생각을 말씀드립니다. 위쪽 cross-attention 기반 layer들는 하드웨어에서는 대단히 효율적으로 구현할 수 있는 부분이고, KV cache가 memory cache reuse가 될 수 있다면 (ex: fit into L3 cache) 이득이 더 증폭될 것으로 생각합니다. 한편, 아래쪽은 좀 더 복잡할 것 같습니다. Sliding window 기반이면 parallel prefill도 어렵지 않고 (blockwise attention을 쓸 수 있으면 더 좋습니다) per-token generation, cache management도 직관적입니다. 반면, Retention 기반이면 chunk-wise로 약간의 recurrence가 생길 수 있어서 prefill stage는 좀 더 오래 걸리고 복잡할 것 같지만, generation stage에서는 load해야 하는 메모리의 양이 엄청나게 적어서 속도가 대단히 빠를 것 같습니다. 요약하면, YOCO는 어떻게 만들더라도 memory bottleneck보다는 compute bottleneck에 훨씬 가까울 것으로 생각합니다.

**Q4**. 최근 양자화를 통한 7B, 8B 모델을 모바일에 serving 하는 형태를 꽤 사용하고 있는 것으로 알고 있습니다. 위와 같이 똑같은 형태로 KV caching을 해결할 수 있다 가정했을 때, 7B 또는 8B 모델을 양자화해서 serving 하는 것과 3B 모델을 그대로 serving 하는 것 발표자 분의 경우 어떤 것을 더 선호하는지 여쭙고 싶습니다.
**A4**. 어려운 질문인데요, 우선 저는 모바일에서는 아직 3B 모델도 quantize하지 않고 serving하는 건 어렵다고 생각합니다. 3B를 float16으로 올리면 파라미터만 6GB RAM을 먹는데, 이건 너무 크다고 보기 때문입니다. 며칠 전 Meta에서 직접 Llama-3.2-1B/3B Quantized model을 공개했는데, 그 경우에도 4-bit weight quantize를 했습니다. 한편, YOCO처럼 만들어서 KV caching으로 인한 memory bottleneck이 상당 부분 해결되어 compute bottleneck이 된다 라고 해도, 저는 속도가 더 빠른 3B 모델 4-bit을 더 선호할 것 같긴 한데, 이건 모바일에서 어느 정도 성능까지 내 주길 바라느냐(간단한 작업 vs 더 복잡한 작업)에 따라 많이 달라질 것 같습니다.

**Q5**. 요즘 KV 캐시의 크기를 줄이기 위해 RetNet, Mamba, 혹은 이번에 발표해주신 YOCO 등의 변형 모델들이 나오고 있는 추세입니다. 이들의 경우 말씀해주신 대로 모델의 크기 면에서는 3B 정도의 크기로 모바일 기기에서 사용될만큼의 크기까지 적절하게 스케일 업이 가능한 것 같은데, 해당 크기의 모델이 다양한 데이터를 학습하는 것도 가능하다고 보실까요? 다시 말해, 기존 Decoder-only LLM들 만큼 많은 양의 데이터를 스케일러블하게 받아들일 수 있을까요? 이에 대한 의견이 궁금합니다.
**A5**. 저는 작은 스케일(~3B)에서는 충분한 데이터가 있다면 "비슷한" 정도로 성능을 낼 수 있다고 믿는 긍정론자입니다 :) 하지만 동시에 Full-context decoder-only Transformer는 항상 upper bound라고 생각하고, 이것보다 좋아지는 건 불가능하다고 보기도 합니다. 그래서 저는 모바일 같은 환경에서 computation-performance trade-off가 특히 중요하게 두드러진다고 생각하고, HW가 따라오기 전까지는 아직 연구자 입장에서 할 수 있는게 꽤 있지 않을까 기대하고 있습니다.
